[Lim Zu Rui] iP

limjeremy496 opened the pull request:
# Vinegar Chatbot

> "Vinegar makes everything a little more refreshing!" – Someone Wise

Vinegar is a simple, efficient, and **fun** chatbot that helps you manage your tasks with ease! Here's why you'll love it:

- **Text-based** for fast input
- **Easy to learn** even for beginners
- **Efficient** task management

## Quick Start Guide

1. **Download** the Vinegar chatbot from [here](#).
2. **Run** the application.
3. **Add tasks** with commands like `todo`, `deadline`, and `event`.
4. **Mark tasks** as done or not done, or even **delete** them if no longer needed. 😄

### Key Features:
- [x] Manage todos
- [x] Add deadlines and events
- [ ] Reminders (coming soon)

If you are a Java developer, Vinegar is also a great project to enhance your **OOP** and **testing** skills! Here's a snippet of the `main` method:

```java
public class Vinegar {
    public static void main(String[] args) {
        new Vinegar("./data/vinegar.txt").run();
    }
}
```

Vinegar is built for:

- Flexibility
- Easy customization
- Simplicity

Check out the full documentation [here](#).

**Bold**, *italic*, and ~~strikethrough~~ text formatting make it easy to style your input too! Get started today, and let Vinegar streamline your task management experience. 😉

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

aureliony committed:
build.gradle: Prevent generating a second JAR file

In build.gradle, the dependencies on distZip and/or distTar causes
the shadowJar task to generate a second JAR file for which the
mainClass.set("seedu.duke.Duke") does not take effect.
Hence, this additional JAR file cannot be run.
For this product, there is no need to generate a second JAR file
to begin with.

Let's remove this dependency from the build.gradle to prevent the
shadowJar task from generating the extra JAR file.

limjeremy496 committed:
Rename main class + Add greeting, exit

limjeremy496 committed:
Add support for echo

limjeremy496 committed:
Add support for list

limjeremy496 committed:
Add Task class and support for mark as done/undone

limjeremy496 committed:
Add support for ToDos, Events, Deadlines

limjeremy496 committed:
Add support for Automated Text UI Testing

limjeremy496 committed:
Add support to handle errors

limjeremy496 committed:
Add support for delete

limjeremy496 committed:
Update for input test case

limjeremy496 committed:
Add support for save

limjeremy496 committed:
Merge branch 'branch-Level-7'

* branch-Level-7:
  Add support for save

limjeremy496 committed:
Add support for parsing and formatting date and time in Deadline and Event

limjeremy496 committed:
Merge branch 'branch-Level-8'

* branch-Level-8:
  Add support for parsing and formatting date and time in Deadline and Event

limjeremy496 committed:
Refactor code to adhere to OOP principles: Extracted classes for Ui, Storage, Parser, TaskList, and Command structure

limjeremy496 committed:
Merge branch 'A-MoreOOP'

* A-MoreOOP:
  Refactor code to adhere to OOP principles: Extracted classes for Ui, Storage, Parser, TaskList, and Command structure

limjeremy496 committed:
Refactor: Organise classes into packages (vinegar.ui, vinegar.storage, vinegar.command, vinegar.task)

limjeremy496 committed:
Merge branch 'A-Packages'

* A-Packages:
  Refactor: Organise classes into packages (vinegar.ui, vinegar.storage, vinegar.command, vinegar.task)

limjeremy496 committed:
Handle invalid task number input in DeleteCommand

limjeremy496 committed:
Minor updates to vinegar.txt

limjeremy496 committed:
Merge branch 'add-gradle-support'

* add-gradle-support:
  build.gradle: Prevent generating a second JAR file
  Bump gradle and lib version
  Add Gradle support

limjeremy496 committed:
Add JUnit tests for TaskList, Vinegar, AddEventCommand

limjeremy496 committed:
Handle invalid input in command package

limjeremy496 committed:
Merge branch 'A-Jar'

* A-Jar:
  Handle invalid input in command package

uniqly commented on src/main/java/vinegar/Parser.java:
```diff
@@ -0,0 +1,31 @@
+package vinegar;
+
+import vinegar.command.*;
```
Should this wildcard import be split into [separate import statements](https://se-education.org/guides/conventions/java/intermediate.html#package-and-import-statements)? Perhaps your IDE is [not configured correctly](https://se-education.org/guides/tutorials/intellijCodeStyle.html#tweak-import-order)? I've noticed this in other files too.

uniqly commented on src/main/java/vinegar/command/AddEventCommand.java:
```diff
@@ -0,0 +1,52 @@
+    private LocalDateTime to;
+
+    public AddEventCommand(String[] inputParts) throws VinegarException {
+        Validator.validateParts(inputParts, 2, "Please specify the description and event time.");
+        String[] eventParts = inputParts[1].split(" /from | /to ");
+        Validator.validateParts(eventParts, 3, "Please specify the event time using /from and /to in the format yyyy-MM-dd HH:mm.");
```
Should these lines be split to fit the [120 character line limit](https://se-education.org/guides/tutorials/intellijCodeStyle.html#tweak-import-order)?

uniqly commented on src/main/java/vinegar/task/Deadline.java:
```diff
@@ -0,0 +1,28 @@
+
+public class Deadline extends Task {
+
+    protected LocalDateTime by;
```
Should this instance field `by` be declared **_after_** the static fields such as `inputFormatter` (instead of being _before_ them)?

uniqly commented on src/main/java/vinegar/task/Deadline.java:
```diff
@@ -0,0 +1,28 @@
+public class Deadline extends Task {
+
+    protected LocalDateTime by;
+    private static final DateTimeFormatter inputFormatter = DateTimeFormatter.ofPattern("yyyy-MM-dd HH:mm");
+    private static final DateTimeFormatter outputFormatter = DateTimeFormatter.ofPattern("MMM d yyyy HH:mm");
```
Should the names of these constants be in SCREAMING_SNAKE_CASE?

uniqly commented on src/main/java/vinegar/task/Event.java:
```diff
@@ -0,0 +1,36 @@
+    protected LocalDateTime from;
+    protected LocalDateTime to;
+
+    private static final DateTimeFormatter inputFormatter = DateTimeFormatter.ofPattern("yyyy-MM-dd HH:mm");
+    private static final DateTimeFormatter outputFormatter = DateTimeFormatter.ofPattern("MMM d yyyy HH:mm");
```
Same issue in `Deadline.java`.

uniqly commented on src/main/java/vinegar/storage/Storage.java:
```diff
@@ -0,0 +1,62 @@
+import vinegar.task.Task;
+import vinegar.task.Todo;
+
+import java.io.*;
```
Same issue as in `Parser.java`.

jeremychiaaaaa commented on src/main/java/vinegar/Parser.java:
```diff
@@ -0,0 +1,31 @@
+        String[] inputParts = fullCommand.split(" ", 2);
+        String instruction = inputParts[0].toLowerCase();
+
+        switch (instruction) {
+            case "bye":
+                return new ExitCommand();
+            case "list":
+                return new ListCommand();
+            case "mark":
+                return new MarkCommand(inputParts);
+            case "unmark":
+                return new UnmarkCommand(inputParts);
+            case "todo":
+                return new AddTodoCommand(inputParts);
+            case "deadline":
+                return new AddDeadlineCommand(inputParts);
+            case "event":
+                return new AddEventCommand(inputParts);
+            case "delete":
+                return new DeleteCommand(inputParts);
+            default:
```
Should the case and default statement be on the same level as the switch line?

uniqly replied:
Should the `case` and `default` keywords for the a `switch` statement be on same indentation level? Did you [configure](https://se-education.org/guides/tutorials/intellijCodeStyle.html#tweak-switch-case-style) your IDE correctly?

jeremychiaaaaa commented on src/main/java/vinegar/Validator.java:
```diff
@@ -0,0 +1,18 @@
+        }
+    }
+
+    private static boolean anyPartEmpty(String[] parts) {
```
I like the function name as it explains accurately the purpose of the function!

jeremychiaaaaa commented on src/main/java/vinegar/storage/Storage.java:
```diff
@@ -0,0 +1,62 @@
+    }
+
+    public List<Task> load() throws IOException {
+        List<Task> tasks = new ArrayList<>();
```
I like the accurate use of the plural form for collection type variables!

jeremychiaaaaa commented on src/main/java/vinegar/ui/Ui.java:
```diff
@@ -0,0 +1,78 @@
+    }
+
+    public void showWelcome() {
+        System.out.println("____________________________________________________________");
+        System.out.println("Hello! I'm vinegar.Vinegar\n" + "What Can I do for you?");
+        System.out.println("____________________________________________________________");
```
Should they be on the same indentation ?

limjeremy496 replied:
Can you clarify on this?

jeremychiaaaaa commented on src/test/java/vinegar/VinegarTest.java:
```diff
@@ -0,0 +1,99 @@
+    }
+
+    @Test
+    public void testRunCommand_AddDeadline() throws IOException, VinegarException {
```
I like the accurate use of unitBeingTested_descriptionOfTestInputs_expectedOutcome convention!

jeremychiaaaaa reviewed (commented):
Overall great job in following 'naming' code quality guidelines. I really liked the descriptive method names that helped my understanding of the method on first glance!

uniqly commented on src/main/java/vinegar/storage/Storage.java:
```diff
@@ -0,0 +1,62 @@
+                String description = parts[2];
+                Task task = null;
+
+                switch (type) {
+                    case "T":
+                        task = new Todo(description);
+                        break;
+                    case "D":
+                        task = new Deadline(description, parts[3]);
+                        break;
+                    case "E":
+                        task = new Event(description, parts[3], parts[4]);
+                        break;
+                }
```
Same issue in `Parser.java`.

uniqly commented on src/main/java/vinegar/task/Event.java:
```diff
@@ -0,0 +1,36 @@
+    }
+
+    @Override
+    public String toString() {
+        return "[E]" + super.toString() + " (from: " + from.format(outputFormatter) + " to: "
+                + to.format(outputFormatter) + ")";
+    }
+
+    @Override
+    public String toFileFormat() {
+        return "E | " + (isDone ? "1" : "0") + " | " + description + " | " + from.format(inputFormatter)
+                + " | " + to.format(inputFormatter);
+    }
```
Great job for following the coding convention for line breaks! 👍

uniqly reviewed (commented):
A decent effort to following the [coding standard](https://se-education.org/guides/conventions/java/index.html#layout). However, violations related to default to IDE configuration are observed. In addition, JavaDoc is missing for classes and non-trivial public methods.

limjeremy496 committed:
Add support for javadocs

limjeremy496 committed:
Add support for find

limjeremy496 committed:
Add support for GUI using JavaFX

limjeremy496 committed:
Merge branch 'branch-Level-10'

* branch-Level-10:
  Add support for GUI using JavaFX

limjeremy496 committed:
Integrated JavaFx GUI with bot's Ui class

There are several lines that are now integrated into DialogBox.

Integrating the GUI improves the bot's aesthetics.

As a step towards Ui improvements, let's make use of Scenebuilder. Doing so will
make the subsequent Ui adjustments easier.

limjeremy496 committed:
Add assertions to validate assumptions in Vinegar class

There are several points in the Vinegar class where important assumptions should hold, such as valid input, task initialisation, and command parsing.

Introducing assert statements improves the code quality, ensuring these assumptions are explicitly verified at runtime.

As a step towards improving robustness, let's add assertions in key locations. This will make the program more reliable and help identify issues early in development.

limjeremy496 committed:
Add assertions for file handling and task validation in Storage

Introduced assert statements to ensure the correctness of directory paths, file existence, task line formatting, and task list validity during load and save operations. These checks help catch unexpected issues during file I/O processes.

limjeremy496 committed:
Merge branch 'A-Assertions'

* A-Assertions:
  Add assertions for file handling and task validation in Storage

limjeremy496 committed:
Refactor Storage class for improved code quality

Incorporated Validator class to handle validation logic, reducing method length and improving readability. Applied SLAP to ensure single-responsibility methods and removed deep nesting.

limjeremy496 committed:
Merge branch 'A-CodeQuality'

* A-CodeQuality:
  Refactor Storage class for improved code quality

limjeremy496 committed:
Add help command and load sample data on first run

Implemented help command to provide in-app guidance with a help page showing available commands.

Modified Storage class to load sample tasks (Todo, Deadline, Event) if no task storage file exists, helping new users get started.

limjeremy496 committed:
Update bye command to exit program

limjeremy496 committed:
Merge branch 'C-Help'

* C-Help:
  Update bye command to exit program
  Add help command and load sample data on first run

limjeremy496 committed:
Update README file

limjeremy496 committed:
Merge branch 'A-UserGuide'

* A-UserGuide:
  Update README file

limjeremy496 committed:
Update README.md

Update user guide

limjeremy496 committed:
Update README.md

limjeremy496 committed:
Update README.md

limjeremy496 committed:
Update README.md

limjeremy496 committed:
Update README.md

limjeremy496 committed:
Fix exit command to close window and other bugs

limjeremy496 committed:
Add delete all tasks function

limjeremy496 committed:
Update README.md

limjeremy496 committed:
Update README.md

added release link

limjeremy496 committed:
Update README.md

limjeremy496 committed:
Update README.md
